Nick Proskurin 181.1, task A: HashMap, attempt 1

MCPN opened the pull request:
/shrug
¯\_(ツ)_/¯

MCPN committed:
Create HashMap.h

MCPN committed:
cp -> cap (capacity)

MCPN committed:
tmp -> prev_sz

MCPN committed:
constants received names

MCPN committed:
another constant received a name

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType>>
+class HashMap {
+public:
+    using iterator = typename std::list<std::pair<const KeyType, ValueType>>::iterator;
```
const_iterator и InitList - названия в разном стиле

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+    using const_iterator = typename std::list<std::pair<const KeyType, ValueType>>::const_iterator;
+    using InitList = std::initializer_list<std::pair<const KeyType, ValueType>>;
+
+    HashMap(const Hash &hshr=Hash()): sz(0), cap(START_SIZE), hasher(hshr), items(), table() {
```
= отбивается пробелами

glebshp replied:
и hshr - непонятное название переменной

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+    }
+
+private:
+    size_t MAX_FACTOR = 2, MIN_FACTOR = 8, START_SIZE = 5;
```
сделать public const static и перенести в начало

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+
+    HashMap(const HashMap &other): HashMap(other.hasher) {
+        if (this != &other) {
+            for (auto elem : other) {
```
const auto

glebshp replied:
а где const?

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    HashMap(const InitList &lst, const Hash &hshr=Hash()): HashMap(hshr) {
```
пробелы

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    HashMap(const InitList &lst, const Hash &hshr=Hash()): HashMap(hshr) {
+        for (auto elem : lst) {
```
const

glebshp replied:
и lst не лучшее название

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    HashMap& operator =(const HashMap &other) {
```
пробел

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+
+    void insert(const std::pair<const KeyType, ValueType> &elem) {
+        put(elem);
+        if (MAX_FACTOR * sz > cap) {
```
MAX_FACTOR - это марка косметики, и, если это не product placement, то константу можно назвать более информативно

MCPN replied:
Переименую в MAX_LOAD_FACTOR (аналогично тому, как это сделано в unordered_map) и MIN_LOAD_FACTOR соответственно.

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+        return end();
+    }
+
+    void rehash(bool shrink=false) {
```
Разумнее сделать три метода:
1. Изменение до заданного размера
2. Увеличение
3. Уменьшение

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,220 @@
+    void del(const KeyType &key) {
+        size_t pos = hasher(key) % cap;
+        for (size_t i = 0; i < cap; ++i) {
+            if (table[pos].second == 1 && table[pos].first -> first == key) {
```
-> не отбивают пробелами

glebshp reviewed (changes requested):
Я, возможно, что-то неправильно смотрел, но где итератор по таблице?

MCPN commented:
Разве не достаточно просто определить методы begin() и end()? Специально для таблицы я итератор естественно не писал, вместо этого сослался на итератор списка.

MCPN committed:
changings required by reviewer

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+    void erase(const KeyType &key) {
+        del(key);
+        if (!empty() && MIN_LOAD_FACTOR * sz < cap) {
+            shrink();
```
не то, такая логика стала только менее понятной.
Должно быть:
1. Функция rehash, которая принимает точный размер, в который надо изменить
2. Функция проверки, не надо ли увеличить, функция проверки, не надо ли уменьшить. Либо общая функция, которая возвращает новый размер, до которого надо делать resize.
3. Далее либо предыдущие функции вместе вызываются в местах, где может произойти изменение, но проще сделать общую функцию "проверь и поменяй", которая уже содержит в себе проверку, что надо сделать, и, если надо, делает

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+    void clear() {
+        size_t prev_sz = sz;
+        for (size_t i = 0; i < prev_sz; ++i) {
+            erase(items.begin()->first);
```
Так удалять довольно не эффективно, явно удаление оптом можно реализовать дешевле (с точки зрения времени работы)

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+        size_t prev_sz = sz;
+        for (size_t i = 0; i < prev_sz; ++i) {
+            put(*items.begin());
+            items.erase(items.begin());
```
разумнее не вызывать каждый раз erase, а пройтись по старому списку а в конце удалить его оптом

MCPN replied:
Сделал без erase (не совсем уверен, что то, что требовалось), стало работать чуть дольше. На всякий случай залью, но вообще странно.

glebshp reviewed (changes requested):
Мелочи + функции проверки на resize и собственно resize сейчас организованы странно

MCPN committed:
auto -> const auto

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+
+    HashMap(const HashMap &other): HashMap(other.hasher) {
+        if (this != &other) {
+            for (const auto elem : other) {
```
const auto&

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    HashMap(const init_list &initializer, const Hash &hash_function = Hash()): HashMap(hash_function) {
+        for (const auto elem : initializer) {
```
const auto&

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+        if (this != &other) {
+            hasher = other.hasher;
+            clear();
+            for (const auto elem : other) {
```
const auto& здесь и далее, не буду уже отмечать

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    void shrink() {
+        cap /= SHRINK_FACTOR;
```
хорошо бы не уводить в меньше чем стартовый размер

glebshp commented on HashMap.h:
```diff
@@ -0,0 +1,226 @@
+
+    void insert(const std::pair<const KeyType, ValueType> &elem) {
+        put(elem);
+        if (MAX_LOAD_FACTOR * sz > cap) {
```
А теперь ещё можно засунуть в один метод (resize_if_needed например) проверки размеров, вызов expand/shrink и resize

glebshp reviewed (changes requested):
Разные мелочи + я всё ещё хочу clear без erase

MCPN committed:
const auto -> const &auto

MCPN committed:
changes required by reviewer

clear(): no erase is called
rehash() -> check_and_rehash(): all conditions are now in function, no erase is called
+ some minor fixes

glebshp reviewed (approved):
Норм
